impl/maven-core/src/main/java/org/apache/maven/project/DefaultProjectBuilder.java

```diff
@@ -492,10 +491,14 @@ List<ProjectBuildingResult> build(List<File> pomFiles, boolean recursive) throws
                         .findAny()
                         .orElse(null);
                 if (cycle != null) {
-                    throw new RuntimeException(new ProjectCycleException(
+                    final CycleDetectedException cde = (CycleDetectedException) cycle.getException();
```

Comment:
The wrapping still smells like something's not quite right. The code is quite complex, and it might be simpler without the lambda, but either way we really should be able to just throw the CycleDetectedException. This is not a public method so we can easily add that to the method signature. It's also possible CycleDetectedException should be a subclass of ProjectBuildingException.